docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+	Status ExternalServiceProviderStatus
+}
+
+type ExternalServiceProviderSpec struct {
+	Host string
+	Ports []ExternalServiceProviderPort
+	LivenessProbe *Probe
+	ReadinessProbe *Probe
+}
+
+type ExternalServiceProviderPort struct {
+	Name string // optional - for mapping from ServicePort
+	Protocol Protocol // TCP, UDP
+	Port int
+}
+
+type ExternalServiceProviderStatus struct {
+	Conditions []ExternalServiceProviderCondition
+}
+
+type ExternalServiceProviderCondition struct {
+	Type ExternalServiceProviderConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+
+type ExternalServiceProviderList struct {
+	TypeMeta
+	ListMeta
+	Items []ExternalServiceProvider
+}
+```
+
+```
+type ServiceStatus struct {
+	...
+	Conditions []ServiceCondition
+}
+
+type ServiceCondition struct {
+	Type ServiceConditionType // Readiness, Liveness
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+```
+
+
+### Kubectl (CLI)
+
+Ready Service Endpoints can be queried with the existing `kubectl get endpoints <service-name>`.
+
+Add a `kubectl get externalserviceproviders` (`kubectl get esp` for short) command to list the external service provider instances, their address/port/name and their last known conditions. Components can be retrived with a component specific label, like `type=component` or `component=apiserver`.
+
+
+### Storage
+
+Store ExternalServiceProviders spec/status in etcd, with an apiserver storage wrapper.
+
+
+### Status Updates
+
+Add an external-service-provider-controller that regularly probes registered providers and updates their stored last known status.
+
+Add self-registration to the components that uses the new ExternalServiceProviders API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new ExternalServiceProviders API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new ExternalServiceProviders API (Update call) when the component errors/panics.
+
+Use a TCP probe (of `/healthz` or `/healthz/ping`) to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints with sub-resources to allow them to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of immediate dependencies (e.g. etcd or other databases).
+
+
+## Example
+
+The following terminal commands are an example of how the `kubectl get externalserviceproviders` command might work.
+
+TODO
+
+## Alternate Naming
+
+ExternalServiceProvider is a bit verbose for a top-level API resource name. Here are some other naming options for the same concept.
```

Comment:
I would also like to make the terminology and style consistent. But consistency is only important because it serves usability and maintainability. Being consistent with something that is poorly understood will only serve to make it consistently confusing. You need a story you can tell about why it is the way it is. 

I don't mind entertaining alternative names, that's why I posted a list of alternatives in the first place, because I was unsatisfied with the naming (after being pressured to abandon the Component naming that seemed so clear and concise for something more broad and ambiguous). But then we debated the naming for weeks and ExternalServiceProvider was the best we came up with.

I don't agree with your reasoning about "external". By your own admission, `ExternalIP` is the only usage, and I've already provided a story that works for both usages: external means a reference to something not managed by k8s. Kubernetes doesn't manage the load balancers or network routing that an external client uses to connect with a service any more than it manages external service providers. It just keeps a reference to them.

The objection to Service has also already made, and no suitable alternative was proposed. ExternalProvider doesn't say what it provides. ExternalEndpoint is confusing because it implies there are InternalEndpoints and conflicts with the specificity and plurality of Endpoints itself. ExternalPod was proposed too, but Pod implies a collection, which an external reference doesn't have. It'd be nice to use Endpoint, singular, but that's super confusing with Endpoints plural already around. So we compromised on something less than perfect in order to be able to preserve reverse compatibility.